core/src/test/java/com/novoda/merlin/service/ConnectivityChangesForwarderTest.java

```diff
@@ -0,0 +1,146 @@
+package com.novoda.merlin.service;
+
+import com.novoda.merlin.NetworkStatus;
+import com.novoda.merlin.receiver.ConnectivityChangeEvent;
+import com.novoda.merlin.registerable.bind.BindListener;
+import com.novoda.merlin.registerable.connection.ConnectListener;
+import com.novoda.merlin.registerable.disconnection.DisconnectListener;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+
+public class ConnectivityChangesForwarderTest {
+
+    private static final boolean CONNECTED = true;
+    private static final boolean DISCONNECTED = false;
+    private static final String ANY_INFO = "any_info";
+    private static final String ANY_REASON = "any_reason";
+
+    @Rule
+    public MockitoRule mockitoRule = MockitoJUnit.rule();
+
+    @Mock
+    private NetworkStatusRetriever networkStatusRetriever;
+    @Mock
+    private DisconnectListener disconnectListener;
+    @Mock
+    private ConnectListener connectListener;
+    @Mock
+    private BindListener bindListener;
+    @Mock
+    private HostPinger hostPinger;
+
+    private ConnectivityChangesForwarder connectivityChangesForwarder;
+
+    @Before
+    public void setUp() {
+        connectivityChangesForwarder = new ConnectivityChangesForwarder(
+                networkStatusRetriever,
+                disconnectListener,
+                connectListener,
+                bindListener,
+                hostPinger
+        );
+    }
+
+    @Test
+    public void givenPreviousConnectedNetworkStatus_whenNotifyingOfInitialState_thenCallsBindListenerWithNetworkStatus() {
+        givenPreviousNetworkStatusIs(CONNECTED);
+
+        connectivityChangesForwarder.notifyOfInitialNetworkStatus();
+
+        verify(bindListener).onMerlinBind(NetworkStatus.newAvailableInstance());
+    }
+
+    @Test
+    public void givenPreviousDisconnectedNetworkStatus_whenNotifyingOfInitialState_thenCallsBindListenerWithNetworkStatus() {
+        givenPreviousNetworkStatusIs(DISCONNECTED);
+
+        connectivityChangesForwarder.notifyOfInitialNetworkStatus();
+
+        verify(bindListener).onMerlinBind(NetworkStatus.newUnavailableInstance());
+    }
+
+    @Test
+    public void givenNoPreviousNetworkStatus_whenNotifyingOfInitialState_thenCallsBindListenerWithNetworkStatusFromRetriever() {
```

Comment:
Ditto: `givenNetworkWasNotAvailable` and `thenForwardsNetworkAvailableToListener`. I would also extract `NetworkStatus.newAvailableInstance()` as a constant given you expect it to be the same and this is relying on `NetworkStatus` on returning cached instances, which is an assumption we maybe don't want to make as it's a bit too whitebox.